OutOfSchool/OutOfSchool.WebApi/Services/Images/ImageDependentEntityImagesInteractionMediator.cs

```diff
@@ -21,152 +21,168 @@ namespace OutOfSchool.WebApi.Services.Images
+                result.UploadedMultipleResult = await UploadManyImagesProcessAsync(entity, newImages).ConfigureAwait(false);
             }
 
-            return await RemoveImageProcessAsync(entity, imageId).ConfigureAwait(false);
+            Logger.LogTrace("Changing images for the entity was finished");
+            return result;
         }
 
         /// <inheritdoc/>
-        public async Task<ImageUploadingResult> UploadManyImagesAsync(TKey entityId, IList<IFormFile> images)
+        public async Task<Result<string>> AddCoverImageAsync(TEntity entity, IFormFile image)
         {
-            if (images == null || images.Count <= 0)
+            _ = entity ?? throw new ArgumentNullException(nameof(entity));
+            _ = image ?? throw new ArgumentNullException(nameof(image));
+
+            if (!string.IsNullOrEmpty(entity.CoverImageId))
             {
-                return new ImageUploadingResult
-                    { MultipleKeyValueOperationResult = new MultipleKeyValueOperationResult { GeneralResultMessage = ImagesOperationErrorCode.NoGivenImagesError.GetResourceValue() } };
+                return Result<string>.Failed(ImagesOperationErrorCode.ImageAlreadyExist.GetOperationError());
             }
 
-            var entity = await GetEntityWithIncludedImages(entityId).ConfigureAwait(false);
-            if (entity == null)
+            var uploadingCoverImageResult = await ImageService.UploadImageAsync<TEntity>(image).ConfigureAwait(false);
+
+            if (uploadingCoverImageResult.Succeeded)
             {
-                Logger.LogError($"Entity [id = {entityId}] was not got.");
-                return new ImageUploadingResult
-                    { MultipleKeyValueOperationResult = new MultipleKeyValueOperationResult { GeneralResultMessage = ImagesOperationErrorCode.EntityNotFoundError.GetResourceValue() } };
+                entity.CoverImageId = uploadingCoverImageResult.Value;
             }
 
-            return await UploadManyImagesProcessAsync(entity, images).ConfigureAwait(false);
+            return uploadingCoverImageResult;
         }
 
         /// <inheritdoc/>
-        public async Task<ImageRemovingResult> RemoveManyImagesAsync(TKey entityId, IList<string> imageIds)
+        public async Task<OperationResult> RemoveCoverImageAsync(TEntity entity)
         {
-            if (imageIds == null || imageIds.Count <= 0)
+            _ = entity ?? throw new ArgumentNullException(nameof(entity));
+            if (string.IsNullOrEmpty(entity.CoverImageId))
             {
-                return new ImageRemovingResult
-                {
-                    MultipleKeyValueOperationResult = new MultipleKeyValueOperationResult { GeneralResultMessage = ImagesOperationErrorCode.NoGivenImagesError.GetResourceValue() },
-                };
+                throw new InvalidOperationException(@"Image id must be a non empty string");
```

Comment:
Probably better to use `ArgumentException`, as it's an issue with input argument, not a processing code